internal/cmd/gocdk/fswatch.go

```diff
@@ -0,0 +1,205 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"golang.org/x/xerrors"
+)
+
+// waitForDirChange blocks until the directory changes or the context's Done
+// channel is closed.
+func waitForDirChange(ctx context.Context, path string, pollInterval time.Duration) error {
+	tick := time.NewTicker(pollInterval)
+	defer tick.Stop()
+	init, err := snapshotTree(path)
+	if err != nil {
+		return xerrors.Errorf("wait for %s to change: %w", path, err)
+	}
+	for {
+		select {
+		case <-tick.C:
+			same, err := init.sameAs(path)
+			if err != nil {
+				// TODO(light): Log error.
+				continue
+			}
+			if !same {
+				return nil
+			}
+		case <-ctx.Done():
+			return xerrors.Errorf("wait for %s to change: %w", path, ctx.Err())
+		}
+	}
+}
+
+// dirSnapshot is a snapshot of metadata for a directory.
+type dirSnapshot struct {
+	entries []*dirSnapshotEntry // sorted by name
+}
+
+// newDirSnapshot converts a list of os.FileInfo into a directory snapshot.
+// It does not fill in any of the entries' snapshot fields.
+func newDirSnapshot(contents []os.FileInfo) *dirSnapshot {
+	snap := &dirSnapshot{
+		entries: make([]*dirSnapshotEntry, 0, len(contents)),
+	}
+	for _, info := range contents {
+		name := info.Name()
+		if isIgnoredForWatch(name) {
+			continue
+		}
+		snap.entries = append(snap.entries, &dirSnapshotEntry{
+			name:    name,
+			mode:    info.Mode(),
+			modTime: info.ModTime(),
+		})
+	}
+	sort.Slice(snap.entries, func(i, j int) bool {
+		return snap.entries[i].name < snap.entries[j].name
+	})
+	return snap
+}
+
+// snapshotTree recursively snapshots a directory.
+func snapshotTree(path string) (*dirSnapshot, error) {
+	// Unfortunately filepath.Walk does not work well for this because
+	// we need to keep track of the parent snapshot variable to write to.
+
+	type walkNode struct {
+		dst  **dirSnapshot
+		path string
+	}
+
+	// Depth-first search over the directory structure.
+	var root *dirSnapshot
+	stack := []walkNode{{&root, path}}
+	for len(stack) > 0 {
+		// Pop next node off stack.
+		curr := stack[len(stack)-1]
+		stack = stack[:len(stack)-1]
+
+		// Read directory and create snapshot.
+		contents, err := ioutil.ReadDir(curr.path)
+		if err != nil {
+			return nil, xerrors.Errorf("snapshot tree %s: %w", path, err)
+		}
+		snap := newDirSnapshot(contents)
+		*curr.dst = snap
```

Comment:
What is this used for?

Reply:
Added comment to clarify.

Reply:
Still not following what `.dst` is used for. If it's filling it in here, maybe it should be initialized to `nil` somewhere else? Maybe this will go away if you go with `os.Walk`.

Reply:
As you say, obsoleted by using `Walk`.